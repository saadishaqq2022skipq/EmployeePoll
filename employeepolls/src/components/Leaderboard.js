import React, { useEffect } from 'react'
import {TableContainer, Table,TableRow, TableCell, TableHead, TableBody, Avatar} from '@mui/material'
import {connect} from 'react-redux'
//import NavBar from './NavBar'
import { useNavigate } from 'react-router-dom'
import './leaderboard.css'

function Leaderboard({users, sorted, auth}) {
  
    const navigate = useNavigate()

    useEffect(()=>{
        !auth && navigate('/login')
    },[auth, navigate])
  
    return (
        users &&
    <div >
    <div className='leaderboard'>
    <TableContainer className='leaderboardTable'>
      <Table sx={{ maxWidth: 650 }} aria-label="simple table" border='2px'>
        <TableHead>
          <TableRow>
            <TableCell>Users</TableCell>
            <TableCell align="right" data-testid='answered'>Answered</TableCell>
            <TableCell align="right">Created</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {sorted.map((row) => (
            <TableRow
              key={row}
              sx={{ '&:last-child td, &:last-child th': { border: 0 } }}
            >
              <TableCell component="th" scope="row">
                <Avatar src={users[row].avatarURL}/>
                {users[row].name}
              </TableCell>
              <TableCell align="right">{Object.keys(users[row].answers).length}</TableCell>
              <TableCell align="right">{Object.keys(users[row].questions).length}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
    </div>
    </div>
  )
}

const mapStateToProps = ({users, authedUser })=>({
    sorted: Object.keys(users).sort(
        (i,j) => {
            const addA = Object.keys(users[i].answers).length + Object.keys(users[i].questions).length
            const addB = Object.keys(users[j].answers).length + Object.keys(users[j].questions).length
            return addB - addA
        }
    ),
    users: users,
    auth: authedUser
})

export default connect(mapStateToProps)(Leaderboard)