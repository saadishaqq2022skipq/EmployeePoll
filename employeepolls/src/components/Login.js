import React from 'react'
//import {  Button, MenuItem, Select } from '@mui/material';
import  './login.css'
import {useState, useEffect} from 'react'
import {setAuthedUser} from '../action/authedUser'
import {connect} from 'react-redux'
import { useNavigate } from 'react-router-dom';

function Login({dispatch, authUser, users}) {


  const [user, setUser] = useState('Select')
  

  const navigate = useNavigate()

  useEffect(()=>{
    authUser && navigate(-1)
  },[authUser, navigate])

  function handleLogin(){

  
      dispatch(setAuthedUser(user))
      navigate(-1)
    
  }
  return (
    <div className='login'>
        
        <div>
          <h1>Login</h1>
        </div>
        <div className='login_text'>
        

        <caption>Select User</caption>
        <select
           value={user}
           onChange={(e)=>setUser(e.target.value)}
           data-testid= 'dropdown' 
           style={{padding:10}}
         >
            <option>Select User</option>
    {Object.keys(users).map(s =>(
             <option  key={users[s].id} value={users[s].id}>
                {users[s].name}
             </option>
    ))}
    
  </select>

       
      </div>
      <div>
        {  user !== 'Select' && user !== 'Select User' &&
        <button  data-testid='login' style={{padding:10}} onClick={handleLogin}>
           Submit
        </button>
        }
        
      </div>
   

        

    </div>
  )
}

const mapStateToProps = ({authUser, users})=>{
  return{
    authUser,
    users
  }
}

export default connect(mapStateToProps)(Login)