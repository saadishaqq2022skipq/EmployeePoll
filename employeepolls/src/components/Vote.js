import React, { useEffect } from 'react'
import {handleSaveAnswer} from '../action/shared'
import {connect} from 'react-redux'
import {useParams} from 'react-router-dom'
import { Avatar,Card,CardActions,Button,CardContent,Typography } from '@mui/material'
import {useNavigate} from 'react-router-dom'
import './vote.css'

function Vote({dispatch, questions, users, auth}) {
    const {question_id} = useParams()
    const navigate = useNavigate()

    
    useEffect(()=>{
        !auth.id && navigate('/login')
    },[auth, navigate])
    function handleVote(answer){
        const info ={
            authedUser:auth.id,
            qid: question_id,
            answer
        }
        console.log(info)
        dispatch(handleSaveAnswer(info))
    }

    if(!questions[question_id]) return (<h1 data-testid='404'>Error 404 Page not found</h1>)
  else return ( users &&
    <div className='VotePoll'>
        <h1>Poll By {questions[question_id].author} </h1>
        <Avatar src= {users[questions[question_id].author].avatarURL} sx={{ width: 250, height: 250 }}/>
        <h3>Would You Rather</h3>
    <div className='voteCard'>
        <div>
        <Card sx={{ maxWidth: 275 }}>
            <CardContent>
        
                 <Typography variant="p" component="div">
                    {questions[question_id].optionOne['text']}
                 </Typography>
        
            </CardContent>
            <CardActions>
                {
                    Object.keys(auth['answers']).includes(question_id) ? 
                    (auth['answers'][question_id]==='optionOne'? (<Button disabled fullWidth variant='outlined' color='success'>Voted</Button>)
                    :(<Button disabled fullWidth variant='outlined' color='success'>Click</Button>)):
                    (<Button fullWidth variant='outlined' color='success' onClick={()=>handleVote('optionOne')}>Click</Button>)
                }
                
            </CardActions>
         </Card>
         <p>Total votes: {questions[question_id].optionOne['votes'].length} ({100*questions[question_id].optionOne['votes'].length/(questions[question_id].optionOne['votes'].length+questions[question_id].optionTwo['votes'].length)}%)</p>
        </div>

        <div>
        
         <Card sx={{ maxWidth: 275 }}>
            <CardContent>
        
                 <Typography variant="p" component="div">
                    {questions[question_id].optionTwo['text']}
                 </Typography>
        
            </CardContent>
            <CardActions>
                {
                    Object.keys(auth['answers']).includes(question_id) ? 
                    (auth['answers'][question_id]==='optionTwo'? (<Button disabled fullWidth variant='outlined' color='success'>Voted</Button>)
                    :(<Button disabled fullWidth variant='outlined' color='success'>Click</Button>)):
                    (<Button fullWidth variant='outlined' color='success' onClick={()=>handleVote('optionTwo')}>Click</Button>)
                }
                
            </CardActions>
         </Card>
         <p>Total votes: {questions[question_id].optionTwo['votes'].length} ({100*questions[question_id].optionTwo['votes'].length/(questions[question_id].optionOne['votes'].length+questions[question_id].optionTwo['votes'].length)}%)</p>
        </div>
        </div>

    </div>
  )
}

const mapStateToProps = ({users, authedUser, questions}) => ({
    auth: authedUser ? users[authedUser]: {id: null},
    questions,
    users
})

export default connect(mapStateToProps)(Vote)