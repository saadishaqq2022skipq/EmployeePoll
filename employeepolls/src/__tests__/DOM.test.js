import React from 'react';
import {fireEvent, render} from '@testing-library/react'
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import Leaderboard from '../components/Leaderboard'
import Navbar from '../components/Navbar';
import Vote from '../components/Vote';
import {store} from '../index'
import Login from '../components/Login';



describe('Testing DOM elements', () => {
    it('will check presence of element', () => {
    const component =  render(
        <MemoryRouter>
            <Provider store={store}>
                <Leaderboard />
            </Provider>
         </MemoryRouter>
        );
        const column = component.getByTestId('answered')
        expect(column).toBeInTheDocument()
    })

    it('will create create snapshot', () => {
        const component =  render(
            <MemoryRouter>
                <Provider store={store}>
                    <Leaderboard />
                </Provider>
             </MemoryRouter>
            );
            expect(component).toMatchSnapshot();
        })

    it('Test Navbar', () => {
        const component =  render(
            <MemoryRouter>
                <Provider store={store}>
                    <Navbar />
                </Provider>
             </MemoryRouter>
        );
        
        const nav1 = component.getByTestId('home')
        expect(nav1).toBeInTheDocument()

        const nav2 = component.getByTestId('leaderboard')
        expect(nav2).toBeInTheDocument()

        const nav3 = component.getByTestId('new')
        expect(nav3).toBeInTheDocument()

        const img = component.getByTestId('logout')
        expect(img).toBeInTheDocument()
    })

    it('Test Login, login button is not present', () => {
        const component =  render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login />
                </Provider>
             </MemoryRouter>
        ); 
        const drop = component.getByTestId('dropdown')
        expect(drop).toBeInTheDocument()


        const login = component.queryAllByTestId('login')
        expect(login.length).toEqual(0)    
    })

    it('testing fireEvent on Login, login button is present', () => {
        const component =  render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login />
                </Provider>
             </MemoryRouter>
        ); 
        const drop = component.getByTestId('dropdown')
        fireEvent.change(drop, {target : {value: 'tylermcginnis'}});
        const login = component.getByTestId('login')
        expect(login).toBeInTheDocument()
    })

    it('testing 404 on vote when no qid is present', () => {
        const component =  render(
            <MemoryRouter>
                <Provider store={store}>
                    <Vote />
                </Provider>
             </MemoryRouter>
        ); 
        const err = component.getByTestId('404')
        expect(err).toBeInTheDocument()
    })
});