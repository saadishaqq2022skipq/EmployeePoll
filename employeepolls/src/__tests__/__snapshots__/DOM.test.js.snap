// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing DOM elements will create create snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="leaderboard"
        >
          <div
            class="MuiTableContainer-root leaderboardTable css-rorn0c-MuiTableContainer-root"
          >
            <table
              aria-label="simple table"
              border="2px"
              class="MuiTable-root css-1e30f4u-MuiTable-root"
            >
              <thead
                class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
              >
                <tr
                  class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
                >
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                    scope="col"
                  >
                    Users
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                    data-testid="answered"
                    scope="col"
                  >
                    Answered
                  </th>
                  <th
                    class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                    scope="col"
                  >
                    Created
                  </th>
                </tr>
              </thead>
              <tbody
                class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
              />
            </table>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="leaderboard"
      >
        <div
          class="MuiTableContainer-root leaderboardTable css-rorn0c-MuiTableContainer-root"
        >
          <table
            aria-label="simple table"
            border="2px"
            class="MuiTable-root css-1e30f4u-MuiTable-root"
          >
            <thead
              class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                  scope="col"
                >
                  Users
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                  data-testid="answered"
                  scope="col"
                >
                  Answered
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                  scope="col"
                >
                  Created
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            />
          </table>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
